pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/inv/RangeIndexCreator.java

```diff
@@ -0,0 +1,506 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.segment.creator.impl.inv;
+
+import com.google.common.base.Preconditions;
+import it.unimi.dsi.fastutil.Arrays;
+import it.unimi.dsi.fastutil.Swapper;
+import it.unimi.dsi.fastutil.ints.IntComparator;
+import java.io.BufferedOutputStream;
+import java.io.Closeable;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.io.FileUtils;
+import org.apache.pinot.core.query.utils.Pair;
+import org.apache.pinot.core.segment.creator.InvertedIndexCreator;
+import org.apache.pinot.core.segment.creator.impl.SegmentColumnarIndexCreator;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.google.common.base.Charsets.UTF_8;
+import static org.apache.pinot.core.segment.creator.impl.V1Constants.Indexes.BITMAP_RANGE_INDEX_FILE_EXTENSION;
+
+
+/**
+ * Implementation of {@link InvertedIndexCreator} that uses off-heap memory.
+ * <p>We use 2 passes to create the range index.
+ * <ul>
+ *
+ *   <li>
+ *     A
+ *   </li>
+ *   <li>
+ *     In the first pass (adding values phase), when add() method is called, store the raw values into the
+ *     value buffer (for multi-valued column we flatten the values).
+ *     We also store the corresponding docId in docIdBuffer which will be sorted in the next phase based on the value in valueBuffer.
+ *
+ *   </li>
+ *   <li>
+ *     In the second pass (processing values phase), when seal() method is called, we sort the docIdBuffer based on the value in valueBuffer.
+ *     We then iterate over the sorted docIdBuffer and create ranges such that each range comprises of _numDocsPerRange.
+ *     While
+ *   </li>
+ * </ul>
+ */
+public final class RangeIndexCreator implements InvertedIndexCreator {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(RangeIndexCreator.class);
+
+  //This will dump the content of temp buffers and ranges
+  private static final boolean TRACE = false;
+
+  private static final int RANGE_INDEX_VERSION = 1;
+
+  private static final int DEFAULT_NUM_RANGES = 20;
+
+  private static final String VALUE_BUFFER_SUFFIX = "val.buf";
+  private static final String DOC_ID_VALUE_BUFFER_SUFFIX = ".doc.id.buf";
+
+  private final File _rangeIndexFile;
+
+  private final File _tempValueBufferFile;
+  private PinotDataBuffer _tempValueBuffer;
+  private NumberValueBuffer _numberValueBuffer;
+
+  private final File _tempDocIdBufferFile;
+  private PinotDataBuffer _docIdValueBuffer;
```

Comment:
Let's rename _docIdValueBuffer to _docIdBuffer and _docIdBuffer to _docIdBufferWrapperForSorting